Update Container Runtime Interface to use enumerated namespace modes

verb opened the pull request:
**What this PR does / why we need it**: This updates the CRI as described in the [Shared PID Namespace](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/pod-pid-namespace.md#container-runtime-interface-changes) proposal. This change to the alpha API is not backwards compatible: implementations of the CRI will need to update to the new API version.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
WIP #1615

**Special notes for your reviewer**:
/assign @yujuhong 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
[action-required] The Container Runtime Interface (CRI) version has increased from v1alpha1 to v1alpha2. Runtimes implementing the CRI will need to update to the new version, which configures container namespaces using an enumeration rather than booleans.
```

verb committed:
Switch CRI NamespaceOption from bools to enums

verb committed:
Update kubelet for enumerated CRI namespaces

This adds support to both the Generic Runtime Manager and the
dockershim for the CRI's enumerated namespaces.

yujuhong commented:
/cc @kubernetes/sig-node-api-reviews @Random-Liu @mrunalp @resouer

resouer commented:
Since this is a break change (most CRI shims should be impacted), what's the rollout plan and target releases may I ask? (Or has been mentioned somewhere else?)

resouer commented on pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto:
```diff
@@ -174,14 +174,25 @@ message Mount {
+    POD       = 0;
+    CONTAINER = 1;
+    NODE      = 2;
+}
```
These terms are a little bit different from `host_xxx` pattern we used before. So mind to add some comments for them? Like: what is `NamespaceMode=NODE` means? (Seems to be `host_network=true`)

yujuhong replied:
+1 for adding more comments.

feiskyer replied:
+1 of adding comments

maybe rename `NODE` to `HOST`? which I think is consistent with hostNetwork mode

verb replied:
Yup, you're all right, the comments were insufficient. I've updated them now, please take another look.

@feiskyer I chose `NODE` because it's well defined by Kubernetes and I thought we should use the Kubernetes terminology where possible. It does create an inconsistency with v1.Pod (which itself uses host & node inconsistently) but I expect that a v2.Pod would replace the namespace booleans with an enum matching this one.

yujuhong commented:
> Since this is a break change (most CRI shims should be impacted), looks like we are missing rollout plan and target releases? (Or has been mentioned somewhere else?)

@resouer the plan is to target this for v1.10, and the [proposal](https://github.com/kubernetes/community/pull/1048) has been reviewed and approved by sig-node.
This is indeed a breaking change, but I think it'd be better to make this change (to make the API cleaner) since CRI is not completely stable yet. We also have other API breaking changes going out this release (e.g., reopen the log file), so bundle them together is also good.

I understand that this adds more burden to the CRI shim maintainers. As a community we will be working on stabilizing and versioning the API soon. Before that, I think we can be more tolerant to breaking changes and impose that the CRI versions used by kubelet and the shim have to match exactly.

verb commented on pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto:
```diff
@@ -178,14 +178,39 @@ message Mount {
+    // Network namespace for this container/sandbox.
+    // Note: There is currently no way to set CONTAINER scoped network in the Kubernetes API.
+    // Namespaces currently set by the kubelet: POD, NODE
+    NamespaceMode network = 1;
```
@yujuhong I don't know enough about proto wire format to know if changing these types will cause an error on deserialize (which is what I want) or garbage bools. Any ideas?

yujuhong replied:
I think it might not break, but it'd be good to verify the behavior either way.

feiskyer replied:
It breaks since the filed name/type are changed with the same id. If you add those with new ids (e.g. 4,5,6), then it won't break.

verb replied:
Researched a bit here and tested using the remote dockershim. It turns out that [proto3 fields are compatible across types](https://developers.google.com/protocol-buffers/docs/proto3#updating), including bools & enums, so we can't just change the type. The behavior appears to be that the enum will be cast to a bool resulting in lots of containers getting host namespaces.

Instead we can renumber the fields. The old bools will get default false values. The effect is that host network/pid/IPC will stop working until the runtimes update their proto.

yujuhong replied:
I don't now if there is a cleaner way to resolve this short of kubelet performing a version check (which will come later).
Defaulting to POD namespace mode does sound safer.

/cc @Random-Liu @mrunalp

verb replied:
@yujuhong, we could also do something like this:

```
message NamespaceOption {                                                                                                                                                        
    // Prior to using NamespaceMode, NamespaceOption used bools to specify namespace.                                                                                            
    // This unused field is type-incomptabile with the previous bool in order to force                                                                                           
    // an error when a runtime implements the outdated interface.                                                                                                                
    float sentinel = 1;                                                                                                                                                          
    reserved 2 to 3;   
...
```

Then the kubelet would always set `sentinel` to a non-zero which will cause a grpc error and no containers will start:

```
I0204 22:14:02.194405   11572 kuberuntime_manager.go:579] SyncPod received new pod "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)", will create a sandbox for it       
I0204 22:14:02.194427   11572 kuberuntime_manager.go:588] Stopping PodSandbox for "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)", will start new one                  
I0204 22:14:02.194449   11572 kuberuntime_manager.go:640] Creating sandbox for pod "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)"                                     
E0204 22:14:02.195188   11572 remote_runtime.go:92] RunPodSandbox from runtime service failed: rpc error: code = Internal desc = grpc: error unmarshalling request: proto: wrong 
wireType = 5 for field HostNetwork                                                                                                                                               
E0204 22:14:02.195262   11572 kuberuntime_sandbox.go:54] CreatePodSandbox for pod "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)" failed: rpc error: code = Internal de
sc = grpc: error unmarshalling request: proto: wrong wireType = 5 for field HostNetwork                                                                                          
E0204 22:14:02.195281   11572 kuberuntime_manager.go:646] createPodSandbox for pod "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)" failed: rpc error: code = Internal d
esc = grpc: error unmarshalling request: proto: wrong wireType = 5 for field HostNetwork                                                                                         
E0204 22:14:02.195340   11572 pod_workers.go:186] Error syncing pod 49c4f606-09f0-11e8-b5dc-42010af00002 ("debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)"), skipping: f
ailed to "CreatePodSandbox" for "debugtest_default(49c4f606-09f0-11e8-b5dc-42010af00002)" with CreatePodSandboxError: "CreatePodSandbox for pod \"debugtest_default(49c4f606-09f0
-11e8-b5dc-42010af00002)\" failed: rpc error: code = Internal desc = grpc: error unmarshalling request: proto: wrong wireType = 5 for field HostNetwork"                         
I0204 22:14:02.195449   11572 server.go:423] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"debugtest", UID:"49c4f606-09f0-11e8-b5dc-42010af00002", APIVersion:"
v1", ResourceVersion:"271", FieldPath:""}): type: 'Warning' reason: 'FailedCreatePodSandBox' Failed create pod sandbox: rpc error: code = Internal desc = grpc: error unmarshalli
ng request: proto: wrong wireType = 5 for field HostNetwork 
```

tallclair commented:
Shouldn't a breaking change increment the API version? I.e. v1alpha2? Or is that not needed since the wire format is backwards compatible?

verb commented:
@tallclair That's a good point, but I wasn't sure since CRI doesn't use any of the api-machinery and I don't think the version numbers are referenced in the protocol, so I'm not sure we'll gain anything. There's no protocol negotiation, the kubelet will just use the one proto it knows, and we don't want to support the old format going forward.

@yujuhong has plans for versioning and compatibility requirements once the the API stabilizes.

verb commented:
/retest

verb commented:
/retest

verb committed:
Increment CRI version from v1alpha1 to v1alpha2

This also incorporates the version string into the package name so
that incompatibile versions will fail to connect.

Arbitrary choices:
- The proto3 package name is runtime.v1alpha2. The proto compiler
  normally translates this to a go package of "runtime_v1alpha2", but
  I renamed it to "v1alpha2" for consistency with existing packages.
- kubelet/apis/cri is used as "internalapi". I left it alone and put the
  public "runtimeapi" in kubelet/apis/cri/runtime.

yujuhong commented:
> @yujuhong has plans for versioning and compatibility requirements once the the API stabilizes.

This is mentioned in the sig-node meeting today. I think it's okay to bump the CRI version to v1alpha2 in a separate PR.

verb commented:
@tallclair you were totally right and bumping the version worked much better. Thanks for chiming in!

yujuhong commented:
@verb you'd need to change pkg/kubelet/apis/cri/services.go as well.

yujuhong commented:
/cc @Random-Liu @mrunalp @resouer 

This PR bumps the CRI version from v1alpha1 to v1alpha2 since there are breaking changes. Please comment if you have concerns.

feiskyer commented:
> This PR bumps the CRI version from v1alpha1 to v1alpha2 since there are breaking changes. Please comment if you have concerns.

LGTM

verb commented:
> @verb you'd need to change pkg/kubelet/apis/cri/services.go as well.

@yujuhong currently apis/cri/services.go isn't versioned with the runtime and is instead referred to in code as "internalapi". Would you like it to be? It would simplify things.

yujuhong commented:
/lgtm

yujuhong commented:
/assign @dchen1107 
to approve /hack changes

dchen1107 commented:
/lgtm 

/approve

Only reviewed the commit: Update kubelet for enumerated CRI namespaces ...
